File: algorithms/medium/minimum_average_difference.cpp
class Solution {
public:
  int minimumAverageDifference(vector<int> &nums) {
    long long right = 0;
    for (auto n : nums) {
      right += n;
    }
    long long left = 0;
    int ans = 0;
    long long min_avg = LONG_LONG_MAX;
    for (int i = 0; i < nums.size(); ++i) {
      left += nums[i];
      right -= nums[i];
      long long left_part = left / (i + 1);
      long long right_part =
          i == nums.size() - 1 ? 0 : right / (nums.size() - i - 1);
      long long diff = abs(left_part - right_part);
      if (min_avg > diff) {
        ans = i;
        min_avg = diff;
      }
    }
    return ans;
  }
};

File: algorithms/medium/minimum_average_difference.ex
defmodule Solution do
  @spec minimum_average_difference(nums :: [integer]) :: integer
  def minimum_average_difference(nums) do
    right = Enum.sum(nums)
    len = length(nums)

    nums
    |> Enum.reduce({0, 0, right, 0, right}, fn n, {ans, i, max_avg, left, right} ->
      left = left + n
      right = right - n
      left_part = div(left, i + 1)
      right_part = if i == len - 1, do: 0, else: div(right, len - i - 1)
      avg = abs(left_part - right_part)

      if max_avg > avg do
        {i, i + 1, avg, left, right}
      else
        {ans, i + 1, max_avg, left, right}
      end
    end)
    |> elem(0)
  end
end
